File: daily/450-Delete-node-in-a-bst.py
class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        def balanceBST(leftNode, rightNode):
            if not rightNode:
                return leftNode
            rightNode.left = balanceBST(leftNode, rightNode.left)
            return rightNode

        if not root:
            return None

        if root.val == key:
            return balanceBST(root.left, root.right)

        if root.val > key:
            root.left = self.deleteNode(root.left, key)
        elif root.val < key:
            root.right = self.deleteNode(root.right, key)
        return root     

File: daily/Binary Tree/450-Delete-node-in-a-bst.go
package daily

func deleteNode(root *TreeNode, key int) *TreeNode {
	if root == nil {
		return nil
	}

	if root.Val == key {
		return balanceNode(root.Left, root.Right)
	}

	if root.Val > key {
		root.Left = deleteNode(root.Left, key)
	} else if root.Val < key {
		root.Right = deleteNode(root.Right, key)
	}
	return root
}

func balanceNode(leftNode *TreeNode, rightNode *TreeNode) *TreeNode {
	if rightNode == nil {
		return leftNode
	}
	rightNode.Left = balanceNode(leftNode, rightNode.Left)
	return rightNode
}
